### PR TITLE
csi: improve plugin error messages and volume validation

### LINKED ISSUE #7424: CSI: internal plugin errors aren't exposed to operator
If a plugin has an internal error, we log a gRPC error code but that's the only information that we can get according to the CSI specification. An example I encountered was when I stopped a job but the EC2 IAM instance role did not have `DetachVolume` permissions; the job stopped but the EBS volume was still attached to the EC2 instance.

When the `ControllerUnpublishVolume` was called, the client logs show the following:

```
2020-03-22T14:24:57.238Z [WARN] client.csi_client: finished client
unary call: plugin.name=aws-ebs0 plugin.type=controller
grpc.code=Internal duration=161.987586ms
grpc.service=csi.v1.Controller grpc.method=ControllerUnpublishVolume
```

But if we look at the controller plugin's alloc logs they show the real problem:

```
E0322 14:24:57.237631 1 driver.go:109] GRPC error: rpc error: code =
Internal desc = Could not detach volume "vol-09b7801ac621c83f5" from
node "i-0aa3ac7f05e936deb": could not detach volume
"vol-09b7801ac621c83f5" from node "i-0aa3ac7f05e936deb":
UnauthorizedOperation: You are not authorized to perform this
operation. Encoded authorization failure message: [REDACTED]
```

(logs redacted and line-broken for readability)

There's two things to fix here:
* we should retry plugin RPCs (#6863) and then make much louder error messages when they fail.
* we should document that some of the internal behaviors of plugins aren't visible to Nomad and that operators should check the plugin logs when debugging plugin failures.

### PR DESCRIPTION
**Pull request by tgross**
Fixes #7424

Some CSI plugins don't return much for errors over the gRPC socket above and beyond the bare minimum error codes. This changeset improves the operator experience by unpacking the error codes when available and wrapping the error with some user-friendly direction.

Improving these errors also revealed a bad comparison with `require.Error` when `require.EqualError` should be used in the test code for plugin errors. This defect in turn was hiding a bug in volume validation where we're being overly permissive in allowing mount flags, which is now fixed.

cc @angrycub as a heads up

### CODE REVIEW

**Comment by tgross** [comment]
I just did a quick search over the rest of the code base and it looks like this is the only test file with the `require.Error` defect.

File: plugins/csi/client.go [hunk 1]
```diff
@@ -533,6 +596,22 @@ func (c *client) NodePublishVolume(ctx context.Context, req *NodePublishVolumeRe
...
+		code := status.Code(err)
+		switch code {
+		case codes.NotFound:
+			err = fmt.Errorf("volume %q could not be found: %v", req.ExternalID, err)
+		case codes.AlreadyExists:
+			err = fmt.Errorf(
+				"volume %q is already published at target path %q but with capabilities or a read_only setting incompatible with this request: %v",
+				req.ExternalID, req.TargetPath, err)
+		case codes.FailedPrecondition:
+			err = fmt.Errorf("volume %q is already published on another node and does not haveMULTI_NODE volume capability: %v",
```
Comment (drewbailey) [inline plugins/csi/client.go:609]:
```suggestion
			err = fmt.Errorf("volume %q is already published on another node and does not have MULTI_NODE volume capability: %v",
```

File: plugins/csi/client.go [hunk 2]
```diff
@@ -488,6 +524,23 @@ func (c *client) NodeStageVolume(ctx context.Context, volumeID string, publishCo
...
+		code := status.Code(err)
+		switch code {
+		case codes.NotFound:
+			err = fmt.Errorf("volume %q could not be found: %v", volumeID, err)
+		case codes.AlreadyExists:
+			err = fmt.Errorf(
+				"volume %q is already staged to %q but with incompatible capabilities for this request: %v",
+				volumeID, stagingTargetPath, err)
+		case codes.FailedPrecondition:
+			err = fmt.Errorf("volume %q is already published on another node and does not haveMULTI_NODE volume capability: %v",
```
Comment (drewbailey) [inline plugins/csi/client.go:537]:
```suggestion
			err = fmt.Errorf("volume %q is already published on another node and does not have MULTI_NODE volume capability: %v",
```

File: plugins/csi/client.go (hunk 1 above)
Comment (tgross) [reply plugins/csi/client.go]: whoops, thanks

File: plugins/csi/client.go [hunk 3]
```diff
@@ -269,6 +271,24 @@ func (c *client) ControllerPublishVolume(ctx context.Context, req *ControllerPub
...
+				req.ExternalID, req.NodeID, err)
+		case codes.AlreadyExists:
+			err = fmt.Errorf(
+				"volume %q is already published at node %q but with capabilities or a read_only setting incompatible with this request: %v",
+				req.ExternalID, req.NodeID, err)
+		case codes.ResourceExhausted:
+			err = fmt.Errorf("node %q has reached the maximum allowable number of attached volumes: %v",
+				req.NodeID, err)
+		case codes.FailedPrecondition:
+			err = fmt.Errorf("volume %q is already published on another node and does not haveMULTI_NODE volume capability: %v",
```
Comment (drewbailey) [inline plugins/csi/client.go:287]:
```suggestion
			err = fmt.Errorf("volume %q is already published on another node and does not have MULTI_NODE volume capability: %v",
```

File: plugins/csi/client.go [hunk 4]
```diff
@@ -469,7 +505,7 @@ func (c *client) NodeStageVolume(ctx context.Context, volumeID string, publishCo
 	}
 
 	// These errors should not be returned during production use but exist as aids
-	// during Nomad Development
+	// during Nomad development
```
Comment (langmartin) [inline plugins/csi/client.go:508]: def seem less ominous/sarcastic now 😁

**Comment by tgross** [comment]
Cherry-picked to 0.11.3 branch as b904bf4

**Pull request merged**